### ISSUE TITLE
Channel history search

### ISSUE THREAD

**Issue by dskaggs** [ryanguill/watney#11]
Add the ability to search logged messages (either all logs or channel specific)

`!search #channel <term>`
`!search <term>`

Decide what to do with the output:
- DM the requestor with the results
- Create a Gist

**Comment by ryanguill** [comment]
I also want to try to give a certain number of lines of context before and after the search result

**Comment by dskaggs** [comment]
This feature needs to take the security of the specific group/channel into account when returning results. For instance:
- search hits that are in a private group that the requestor is not a member of should not be shown
- there should be a way to manually disable logging in a channel/group if desired but still have Watney in the channel/group for the other features

**Comment by atuttle** [comment]
I know slack has decent search. Is the issue (and the reason we're even considering this) that the free version of slack we're on doesn't keep enough history?

**Comment by dskaggs** [comment]
yeah, that's the idea. We hit the 10k team-wide message threshold late last week

**Comment by ryanguill** [comment]
I think we respond with a gist, mostly because there might be multiple matches and that sounds like the best way.  I think we do +-5 lines of context around the hit by default but we give a way to adjust the context (so if you need a bit more, etc).  I also think we only allow you to search channels, not groups, and so there shouldnt be any non-public channels.  Right now watney logs all channels and groups it is a part of except for the group designated as the "admin" group.  I'm not against making a way to disable logging on a group (and/or clearing the log) but lets do that as part of a separate issue if we deem in necessary - im thinking it isn't at the moment.
